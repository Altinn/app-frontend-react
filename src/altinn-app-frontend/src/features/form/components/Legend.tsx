/* eslint-disable react/prop-types */
import * as React from 'react';
import { ILabelSettings } from 'src/types';
import { getLanguageFromKey } from 'altinn-shared/utils';
import Description from './Description';
import { HelpTextContainer } from './HelpTextContainer';
import { ILanguage } from 'altinn-shared/types';

export interface IFormLegendProps {
  labelText: React.ReactNode;
  descriptionText: React.ReactNode;
  language: ILanguage;
  required?: boolean;
  labelSettings?: ILabelSettings;
  helpText: React.ReactNode;
  id: string;
}

export default function Legend(props: IFormLegendProps) {
  if (!props.labelText) {
    return null;
  }

  return (
    <>
      <label
        className='a-form-label title-label'
        htmlFor={props.id}
      >
        {props.labelText}
         {/* Mark required fields */}
          {props.required ? (
            <span className='label-optional'>
              {` ${getLanguageFromKey('form_filler.required_label', props.language)}`}
            </span>
          ) : null}
          {/* Mark optional fields only if optionalIndicator===true */}
          {props.labelSettings?.optionalIndicator && !props.required ? (
            <span className='label-optional' data-testid='optional-label'>
              {` (${getLanguageFromKey('general.optional', props.language)})`}
            </span>
          ) : null}
        {props.helpText &&
          <HelpTextContainer
            language={props.language}
            helpText={props.helpText}
          />
        }
      </label>
      {props.descriptionText &&
        <Description
          description={props.descriptionText}
          {...props}
        />
      }
    </>
  );
}
