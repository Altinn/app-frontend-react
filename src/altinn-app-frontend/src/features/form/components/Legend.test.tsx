import * as React from 'react';
import { render } from '@testing-library/react';

import Legend, {  IFormLegendProps } from './Legend';

describe('features > form > components >Legend.tsx', () => {
  let mockId: string;
  let mockLegendText: string;
  let mockHelpText: string;
  let mockDescriptionText: string;
  let mockLanguage: any;
  let requiredMarking: string;
  let optionalMarking: string;

  beforeEach(() => {
    mockId =  'label1';
    mockLegendText = 'label.text';
    mockHelpText = '';
    mockDescriptionText = '';
    requiredMarking = '*';
    optionalMarking = 'valgfri';
    mockLanguage = {
      general: {
        optional: optionalMarking,
      },
      'form_filler': {
        'required_label': requiredMarking,
      },
    };
  });

  test('form/components/Legend.tsx -- should render default', () => {
    const { asFragment } = renderLegendComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  test('form/components/Legend.tsx -- should render required marking when field is required', () => {
    const { queryByText } = renderLegendComponent({ required: true });
    expect(queryByText(requiredMarking)).toBeTruthy();
  });

  test('form/components/Legend.tsx -- should render optional marking when labelSettings.optionalIndicator is true', () => {
    const { queryByTestId } = renderLegendComponent({ labelSettings: { optionalIndicator: true } });
    expect(queryByTestId('optional-label')).toBeTruthy();
  });

  test('form/components/Legend.tsx -- should not render optional marking when required, even if labelSettings.optionalIndicator is true', () => {
    const { queryByTestId } = renderLegendComponent({ labelSettings: { optionalIndicator: true }, required: true });
    expect(queryByTestId('optional-label')).toBeFalsy();
  });

  function renderLegendComponent(props: Partial<IFormLegendProps> = {}) {
    const defaultProps: IFormLegendProps = {
      id: mockId,
      labelText: mockLegendText,
      descriptionText: mockDescriptionText,
      helpText: mockHelpText,
      language: mockLanguage,
      required: false,
      labelSettings: {}
    };

    return render(<Legend {...defaultProps} {...props} />);
  }
});
