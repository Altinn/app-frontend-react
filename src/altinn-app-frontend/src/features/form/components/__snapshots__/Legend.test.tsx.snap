// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features > form > components >Legend.tsx form/components/Legend.tsx -- should render default 1`] = `
<DocumentFragment>
  <label
    class="a-form-label title-label"
    for="label1"
  >
    label.text
  </label>
</DocumentFragment>
`;
